common/Resources/ti.internal/extensions/node/fs.js

```diff
@@ -0,0 +1,1257 @@
+	const wasFileDescriptor = (typeof path === 'number');
+	const fileDescriptor = wasFileDescriptor ? path : fs.openSync(path, options.flag); // use default mode
+
+	const stats = fs.fstatSync(fileDescriptor);
+	const fileSize = stats.size;
+
+	// Create a Ti.Buffer to read into
+	// FIXME: file size may be too large, we should throw an error then!
+	const buffer = Ti.createBuffer({ length: fileSize });
+
+	// Read bytes in loop until we hit bytesRead === size, or we get -1 EOF and something went wrong
+	let bytesRead;
+	let position = 0;
+	do {
+		bytesRead = fs.readSync(fileDescriptor, buffer, position, fileSize - position);
+		// FIXME: catch errors and auto-close if thrown?
+		position += bytesRead;
+	} while (bytesRead > 0 && position < fileSize);
+
+	// fs.closeSync if it was not originally a file descriptor
+	if (!wasFileDescriptor) {
+		fs.closeSync(fileDescriptor);
+	}
+
+	// TODO: trim buffer if we didn't read full size?
+
+	return encodeBuffer(options.encoding, buffer);
+};
+
+// TODO: fs.readlink(path[, options], callback)
+// TODO: fs.readlinkSync(path[, options])
+// TODO: fs.realpath(path[, options], callback)
+// TODO: fs.realpath.native(path[, options], callback)
+// TODO: fs.realpathSync(path[, options])
+// TODO: fs.realpathSync.native(path[, options])
+
+/**
+ * @param {string|Buffer|URL} oldPath source filepath
+ * @param {string|Buffer|URL} newPath destination filepath
+ * @param {errorCallback} callback async callback
+ */
+fs.rename = (oldPath, newPath, callback) => {
+	setTimeout(() => {
+		try {
+			fs.renameSync(oldPath, newPath);
+		} catch (e) {
+			callback(e);
+			return;
+		}
+		callback();
+	}, 1);
+};
+
+/**
+ * @param {string|Buffer|URL} oldPath source filepath
+ * @param {string|Buffer|URL} newPath destination filepath
+ */
+fs.renameSync = (oldPath, newPath) => {
+	const tiFile = getTiFileFromPathLikeValue(oldPath);
+	// src doesn't actually exist?
+	if (!tiFile.exists()) {
+		const err = noSuchFile('rename', oldPath);
+		err.message = `${err.message} -> '${newPath}'`;
+		err.dest = newPath;
+		throw err;
+	}
+
+	const destFile = getTiFileFromPathLikeValue(newPath);
+	if (destFile.isDirectory()) {
+		// dest is a directory that already exists
+		const err = illegalOperationOnADirectory('rename', oldPath);
+		err.message = `${err.message} -> '${newPath}'`;
+		err.dest = newPath;
+		throw err;
+	}
+
+	let tempPath;
+	if (destFile.isFile()) {
+		// destination file exists, we should overwrite
+		// Our APIs will fail if we try, so first let's make a backup copy and delete the the original
+		tempPath = path.join(fs.mkdtempSync(path.join(Ti.Filesystem.tempDirectory, 'rename-')), path.basename(newPath));
+		destFile.move(tempPath);
+	}
+
+	let success = false;
+	try {
+		success = tiFile.move(newPath);
+	} finally {
+		if (tempPath) {
+			// we temporarily copied the existing destination to back it up...
+			if (success) {
+				// move worked, so we can wipe it away whenever...
+				fs.unlink(tempPath, err => {});
```

Comment:
<!--
  0 failure: 
  1 warning:  common/Resources/...
  
  
  DangerID: danger-id-Danger;
  File: common/Resources/ti.internal/extensions/node/fs.js;
  Line: 916;
-->

- :warning: common/Resources/ti.internal/extensions/node/fs.js line 916 – 'err' is defined but never used. Allowed unused args must match /^_.+/. (no-unused-vars)